### ISSUE TITLE
About Experiments on Laplacian Regularizer

### ISSUE THREAD

**Issue by lywu18** [yang-han/P-reg#1]
When I run the code with laplacian regularizer, I found it didn't perform as bad as your paper represented, but did better than P-reg. I was wondering if I made something wrong with the parameter setting. Can I know more details about your experiments on laplacian regularizer? I am really confused now why it has such a good performance! Really appreciate it if you reply!!!

**Comment by yang-han** [comment]
Hi, could you provide more details about your experiment setting, e.g., which dataset, model and what hyperparameters, also what the accuracy you obtained?

**Comment by yang-han** [comment]
>  Can I know more details about your experiments on laplacian regularizer?

To answer your question directly, the regularization factor µ is determined by grid search using the validation accuracy.
The search space for µ is 20 values evenly chosen from [0, 1], and the µ remains the same for each cell (model × dataset).

**Comment by lywu18** [comment]
> 

Thanks for your reply! I made the experiment on GAT × Cora. When I used laplacian regularizer in the loss and set mu=0.45, the accuracy is 83.19% while the result using P-reg is 82.23%. My results are all a little lower than your paper represents, which is not important. So I just want to know why the laplacian regularizer made such a good performance on the model.

**Comment by yang-han** [comment]
1. Did you run multiple times (e.g., 10 times) with different random seeds and take the average to exclude the influence of randomness?
2. Not sure whether Laplacian with mu=0.45 is the one with the highest validation accuracy. We can only take the mu with the highest validation accuracy and use it to evaluate the test accuracy, because we are supposed to not able to see the test set labels during training.

Given you description, I think the most possible reason is 1. Could you please check the above two reasons and retry?

**Comment by lywu18** [comment]
I am sure I run multiple times and took the average to get the results. I think maybe I did not describe my question clearly. Your paper represented that laplacian had no contribute to the accuracy while P-reg had a good performance. So no matter whether it is the highest validation accuracy,  the result is so confusing. Thank you for taking your time!

**Comment by yang-han** [comment]
Hi, I rerun my experiments again and found everything is consistent with the paper reported.
For your reference:

Loss with Laplacian Reg:
```python
from loss import laplacian_reg
def adloss(model, data, mask, mu):
    logits_1 = model(data.x, data.edge_index)
    output_1 = torch.log_softmax(logits_1, dim=1)
    loss_1 = F.nll_loss(output_1[mask], data.y[mask])

    loss_2 = laplacian_reg(
        data.edge_index, logits_1
    )
    loss = loss_1 + mu * loss_2

    return loss
```
Parameters you suggested:
```
python main.py --num_splits 5 --num_seeds 5 --dataset cora --model preggat --mu 0.45
```
I got:
```
Namespace(dataset='cora', epochs=10000, lr=0.01, model='preggat', mu=0.45, num_seeds=5, num_splits=5, patience=200, weight_decay=0.0005)
Running using 5 random splits...
seed: 0 Epoch: 58 Train Acc: 0.9642857142857143 Val Acc: 0.8095238095238095 Test Acc: 0.8015267175572519
seed: 1 Epoch: 124 Train Acc: 0.9857142857142858 Val Acc: 0.8 Test Acc: 0.7599660729431722
seed: 2 Epoch: 285 Train Acc: 0.9857142857142858 Val Acc: 0.8238095238095238 Test Acc: 0.765479219677693
seed: 3 Epoch: 295 Train Acc: 0.9714285714285714 Val Acc: 0.819047619047619 Test Acc: 0.7866836301950806
seed: 4 Epoch: 198 Train Acc: 0.9714285714285714 Val Acc: 0.8238095238095238 Test Acc: 0.7841391009329941
seed: 0 Epoch: 19 Train Acc: 0.9285714285714286 Val Acc: 0.8285714285714286 Test Acc: 0.7642069550466497
seed: 1 Epoch: 46 Train Acc: 0.9642857142857143 Val Acc: 0.8285714285714286 Test Acc: 0.7497879558948262
seed: 2 Epoch: 24 Train Acc: 0.9571428571428572 Val Acc: 0.8238095238095238 Test Acc: 0.7493638676844784
seed: 3 Epoch: 246 Train Acc: 0.9928571428571429 Val Acc: 0.8285714285714286 Test Acc: 0.741306191687871
seed: 4 Epoch: 422 Train Acc: 0.9928571428571429 Val Acc: 0.8238095238095238 Test Acc: 0.7519083969465649
seed: 0 Epoch: 37 Train Acc: 0.9857142857142858 Val Acc: 0.8333333333333334 Test Acc: 0.7692960135708228
seed: 1 Epoch: 22 Train Acc: 0.9571428571428572 Val Acc: 0.8666666666666667 Test Acc: 0.8070398642917727
seed: 2 Epoch: 51 Train Acc: 0.9857142857142858 Val Acc: 0.8476190476190476 Test Acc: 0.7765055131467345
seed: 3 Epoch: 52 Train Acc: 0.9857142857142858 Val Acc: 0.861904761904762 Test Acc: 0.7921967769296013
seed: 4 Epoch: 204 Train Acc: 0.9928571428571429 Val Acc: 0.8476190476190476 Test Acc: 0.7692960135708228
seed: 0 Epoch: 204 Train Acc: 0.9785714285714285 Val Acc: 0.819047619047619 Test Acc: 0.7493638676844784
seed: 1 Epoch: 22 Train Acc: 0.9571428571428572 Val Acc: 0.8476190476190476 Test Acc: 0.7854113655640373
seed: 2 Epoch: 28 Train Acc: 0.9642857142857143 Val Acc: 0.819047619047619 Test Acc: 0.73960983884648
seed: 3 Epoch: 24 Train Acc: 0.9642857142857143 Val Acc: 0.8333333333333334 Test Acc: 0.775657336726039
seed: 4 Epoch: 287 Train Acc: 0.9785714285714285 Val Acc: 0.8238095238095238 Test Acc: 0.7595419847328244
seed: 0 Epoch: 23 Train Acc: 0.9642857142857143 Val Acc: 0.8571428571428571 Test Acc: 0.7883799830364716
seed: 1 Epoch: 24 Train Acc: 0.9785714285714285 Val Acc: 0.8476190476190476 Test Acc: 0.779050042408821
seed: 2 Epoch: 24 Train Acc: 0.9785714285714285 Val Acc: 0.8476190476190476 Test Acc: 0.7934690415606446
seed: 3 Epoch: 61 Train Acc: 0.9857142857142858 Val Acc: 0.8476190476190476 Test Acc: 0.7637828668363019
seed: 4 Epoch: 60 Train Acc: 0.9857142857142858 Val Acc: 0.8428571428571429 Test Acc: 0.7599660729431722
Result: Dataset: cora(), Model: <class 'models.PREGGAT'>, Avg. Epochs: 113, Train: 0.9743, Val: 0.8341, Test: 0.77052±0.01828
```

P-reg
```python
def adloss(model, data, mask, mu):
    output_1 = F.softmax(model(data.x, data.edge_index), dim=1)
    loss_1 = F.nll_loss(torch.log(output_1[mask]+EPS), data.y[mask])

    loss_2 = soft_cross_entropy(
        F.softmax(model.propagation(data.x, data.edge_index), dim=1), output_1
    )
    loss = loss_1 + mu * loss_2

    return loss
```
Run with
```
python main.py --num_splits 5 --num_seeds 5 --dataset cora --model preggat --mu 0.45
```
I got
```
Namespace(dataset='cora', epochs=10000, lr=0.01, model='preggat', mu=0.45, num_seeds=5, num_splits=5, patience=200, weight_decay=0.0005)
Running using 5 random splits...
seed: 0 Epoch: 260 Train Acc: 0.9928571428571429 Val Acc: 0.8142857142857143 Test Acc: 0.8316369804919423
seed: 1 Epoch: 171 Train Acc: 0.9857142857142858 Val Acc: 0.8095238095238095 Test Acc: 0.8286683630195081
seed: 2 Epoch: 285 Train Acc: 1.0 Val Acc: 0.8095238095238095 Test Acc: 0.8350296861747244
seed: 3 Epoch: 134 Train Acc: 0.9928571428571429 Val Acc: 0.8095238095238095 Test Acc: 0.8354537743850721
seed: 4 Epoch: 198 Train Acc: 0.9857142857142858 Val Acc: 0.8 Test Acc: 0.8299406276505513
seed: 0 Epoch: 206 Train Acc: 0.9928571428571429 Val Acc: 0.8142857142857143 Test Acc: 0.8341815097540288
seed: 1 Epoch: 613 Train Acc: 1.0 Val Acc: 0.8285714285714286 Test Acc: 0.8413910093299406
seed: 2 Epoch: 21 Train Acc: 0.9357142857142857 Val Acc: 0.819047619047619 Test Acc: 0.7430025445292621
seed: 3 Epoch: 10 Train Acc: 0.9 Val Acc: 0.8142857142857143 Test Acc: 0.7718405428329093
seed: 4 Epoch: 303 Train Acc: 0.9928571428571429 Val Acc: 0.8142857142857143 Test Acc: 0.8379983036471587
seed: 0 Epoch: 271 Train Acc: 1.0 Val Acc: 0.8761904761904762 Test Acc: 0.8409669211195929
seed: 1 Epoch: 303 Train Acc: 1.0 Val Acc: 0.8714285714285714 Test Acc: 0.8384223918575063
seed: 2 Epoch: 324 Train Acc: 1.0 Val Acc: 0.8761904761904762 Test Acc: 0.8413910093299406
seed: 3 Epoch: 42 Train Acc: 0.9714285714285714 Val Acc: 0.8666666666666667 Test Acc: 0.8341815097540288
seed: 4 Epoch: 254 Train Acc: 1.0 Val Acc: 0.8761904761904762 Test Acc: 0.8354537743850721
seed: 0 Epoch: 282 Train Acc: 0.9857142857142858 Val Acc: 0.8857142857142857 Test Acc: 0.8248515691263782
seed: 1 Epoch: 299 Train Acc: 1.0 Val Acc: 0.8857142857142857 Test Acc: 0.8307888040712468
seed: 2 Epoch: 420 Train Acc: 0.9928571428571429 Val Acc: 0.8857142857142857 Test Acc: 0.8295165394402035
seed: 3 Epoch: 258 Train Acc: 0.9928571428571429 Val Acc: 0.8761904761904762 Test Acc: 0.8303647158608991
seed: 4 Epoch: 377 Train Acc: 0.9928571428571429 Val Acc: 0.8666666666666667 Test Acc: 0.8269720101781171
seed: 0 Epoch: 69 Train Acc: 0.9857142857142858 Val Acc: 0.8809523809523809 Test Acc: 0.8269720101781171
seed: 1 Epoch: 208 Train Acc: 0.9857142857142858 Val Acc: 0.8666666666666667 Test Acc: 0.8320610687022901
seed: 2 Epoch: 77 Train Acc: 0.9928571428571429 Val Acc: 0.8714285714285714 Test Acc: 0.8210347752332485
seed: 3 Epoch: 56 Train Acc: 0.9857142857142858 Val Acc: 0.8761904761904762 Test Acc: 0.8197625106022053
seed: 4 Epoch: 68 Train Acc: 0.9857142857142858 Val Acc: 0.8714285714285714 Test Acc: 0.8193384223918575
Result: Dataset: cora(), Model: <class 'models.PREGGAT'>, Avg. Epochs: 220, Train: 0.9860, Val: 0.8507, Test: 0.82565±0.02142
```

Everything is consistent with the results in the paper.

**Comment by lywu18** [comment]
> 
Oh so kind of you! I tried the codes you offered and the results is exactly consistent with the paper. I found that I made some mistakes about the parameters of the laplacian_reg function. Thank you so much for taking so much time to solve my problems, and thank you for this wonderful paper, which helped me learn a lot!!!
